### PR TITLE
Stalfos (en_test) OK and object_sk2 documented

### PR DESCRIPTION
**Pull request by fig02**
Thanks to @EllipticEllipsis for the object documentation.
A good amount of functions are named, but theres still a good amount of docs to do. Will come back to it at a later time.

### CODE REVIEW

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 1]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+    this->actor.home.rot.x = 0;
+
+    if (this->swordState >= STALFOS_SWORD_OFF) {
+        EffectBlure_AddSpace(Effect_GetByIndex(this->effectIndex));
+        this->swordState = STALFOS_SWORD_NOBLUR;
+    }
+
+    this->actor.flags &= ~1;
+
+    if (this->actor.params == 5) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:1522]: In your `StalfosType` enum you go up to type 3, did you miss this? or this is not related to the Stalfos' type?

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 2]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+
+void func_8086318C(EnTest* this, GlobalContext* globalCtx) {
+    if (SkelAnime_Update(&this->skelAnime)) {
+        this->timer = (Rand_ZeroOne() * 10.0f) + 10.0f;
+        this->unk_7C8 = 7;
+        EnTest_SetupAction(this, func_808633E8);
+        BodyBreak_Alloc(&this->bodyBreak, 60, globalCtx);
+    }
+
+    if (((s32)this->skelAnime.curFrame == 10) || ((s32)this->skelAnime.curFrame == 25)) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:1617]: Is the cast necessary to match?

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 3]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+    this->skelAnime.endFrame = 0.0f;
+    EnTest_SetupAction(this, EnTest_Recoil);
+}
+
+void EnTest_Recoil(EnTest* this, GlobalContext* globalCtx) {
+    if (SkelAnime_Update(&this->skelAnime)) {
+        if (Rand_ZeroOne() > 0.7f) {
+            EnTest_SetupIdle(this);
+            this->timer = (Rand_ZeroOne() * 5.0f) + 5.0f;
+        } else if (((globalCtx->gameplayFrames & 1) != 0) && (this->actor.params != STALFOS_CEILING)) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:1637]:
```suggestion
        } else if (((globalCtx->gameplayFrames % 2) != 0) && (this->actor.params != STALFOS_CEILING)) {
```

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 4]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+
+    Animation_MorphToLoop(&this->skelAnime, &gStalfosSidestepAnim, -2.0f);
+    Math_SmoothStepToS(&this->actor.shape.rot.y, this->actor.yawTowardsPlayer, 1, 0xFA0, 1);
+    this->actor.speedXZ = (globalCtx->gameplayFrames & 1) ? -4.0f : 4.0f;
+    this->actor.world.rot.y = this->actor.shape.rot.y + 0x3FFF;
+    this->timer = (Rand_ZeroOne() * 20.0f) + 20.0f;
+    this->unk_7C8 = 0x18;
+    EnTest_SetupAction(this, EnTest_SidestepAgro);
+    this->unk_7EC = 0.0f;
+}
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:1389]: Feel free to ignore this suggestion if you don't like it.

```suggestion
void EnTest_SetupSidestepAgro(EnTest* this, GlobalContext* globalCtx) {
    if (Actor_OtherIsTargeted(globalCtx, &this->actor)) {
        EnTest_SetupSidestepInactive(this);
    } else {
        Animation_MorphToLoop(&this->skelAnime, &gStalfosSidestepAnim, -2.0f);
        Math_SmoothStepToS(&this->actor.shape.rot.y, this->actor.yawTowardsPlayer, 1, 0xFA0, 1);
        this->actor.speedXZ = (globalCtx->gameplayFrames & 1) ? -4.0f : 4.0f;
        this->actor.world.rot.y = this->actor.shape.rot.y + 0x3FFF;
        this->timer = (Rand_ZeroOne() * 20.0f) + 20.0f;
        this->unk_7C8 = 0x18;
        EnTest_SetupAction(this, EnTest_SidestepAgro);
        this->unk_7EC = 0.0f;
    }
}
```

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 5]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+    }
+
+    if ((this->skelAnime.curFrame > 7.0f) && (this->skelAnime.curFrame < 11.0f)) {
+        this->swordState = STALFOS_SWORD_ON;
+    } else {
+        this->swordState = STALFOS_SWORD_OFF;
+    }
+
+    if (SkelAnime_Update(&this->skelAnime)) {
+        if (globalCtx->gameplayFrames % 2) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:941]: Personal preference.
```suggestion
        if (globalCtx->gameplayFrames % 2 != 0) {
```

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 6]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+            if (Actor_IsFacingPlayer(&this->actor, 0x1555)) {
+                if ((this->actor.xzDistToPlayer < 220.0f) && (this->actor.xzDistToPlayer > 160.0f) &&
+                    (Rand_ZeroOne() < 0.3f)) {
+                    if (Actor_IsTargeted(globalCtx, &this->actor)) {
+                        EnTest_SetupJumpslash(this);
+                    } else {
+                        EnTest_SetupSidestepAgro(this, globalCtx);
+                    }
+                } else {
+                    if ((Rand_ZeroOne() > 0.3f)) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:530]:
```suggestion
                    if (Rand_ZeroOne() > 0.3f) {
```

**Review by AngheloAlf — COMMENTED** [review]
What's going on with actor's params and the stalfos' enum?  The usage seems a bit inconsistent.

**Comment by fig02** [comment]
Regarding the type enum, I simply didnt get around to documenting what every type is. I will use the enum for other types with a generic name for now though.

File: src/overlays/actors/ovl_En_Test/z_en_test.c (hunk 2 above)
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:1617]: it is yeah, curFrame is a float

File: src/overlays/actors/ovl_En_Test/z_en_test.c (hunk 4 above)
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:1389]: Both work well. This actor in general has alot of early returns cause of early action changes. This one only happens to look good cause the block beneath it isnt that large. Ill prob leave it as is

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 7]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+            EnTest_SetupJumpUp(this);
+            return true;
+        }
+
+        if ((ABS(yawToProjectile) < 0x2000) || (ABS(yawToProjectile) > 0x6000)) {
+            directionFlag = globalCtx->gameplayFrames % 2;
+
+            if (touchingWall && (wallYawDiff > 0x2000) && (wallYawDiff < 0x6000)) {
+                directionFlag = false;
+            } else if ((touchingWall != 0) && (wallYawDiff < -0x2000) && (wallYawDiff >= -0x5FFF)) {
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:2059]:
```suggestion
            } else if ((touchingWall != 0) && (wallYawDiff < -0x2000) && (wallYawDiff > -0x6000)) {
```

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 8]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+    EnTest* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Test/func_80862DBC.s")
+    Actor_ProcessInitChain(&this->actor, sInitChain);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Test/func_80862E6C.s")
+    SkelAnime_Init(globalCtx, &this->skelAnime, &gStalfosSkel, &gStalfosMiddleGuardAnim, this->jointTable,
+                   this->morphTable, 61);
+    SkelAnime_Init(globalCtx, &this->upperSkelanime, &gStalfosSkel, &gStalfosMiddleGuardAnim, this->upperJointTable,
+                   this->upperMorphTable, 61);
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:271]:
```suggestion
                   this->morphTable, STALFOS_LIMB_MAX);
    SkelAnime_Init(globalCtx, &this->upperSkelanime, &gStalfosSkel, &gStalfosMiddleGuardAnim, this->upperJointTable,
                   this->upperMorphTable, STALFOS_LIMB_MAX);
```

**Review by Dragorn421 — CHANGES_REQUESTED** [review]
Only minor "style" things, didn't look at much of the doc yet

(Anyone wants to start a grammar war about "agro" vs "aggro"? (I'm all for "aggro" as in aggressive but apparently people can't spell so both coexist))

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 9]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+    /* Unblockable   */ DMG_ENTRY(0, STALFOS_DMGEFF_NORMAL),
+    /* Hammer jump   */ DMG_ENTRY(4, STALFOS_DMGEFF_NORMAL),
+    /* Unknown 2     */ DMG_ENTRY(0, STALFOS_DMGEFF_NORMAL),
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Test/func_80862650.s")
+static InitChainEntry sInitChain[] = {
+    ICHAIN_S8(naviEnemyId, 27, ICHAIN_CONTINUE),      ICHAIN_F32(targetArrowOffset, 500, ICHAIN_CONTINUE),
+    ICHAIN_VEC3F_DIV1000(scale, 15, ICHAIN_CONTINUE), ICHAIN_F32(scale.y, 0, ICHAIN_CONTINUE),
+    ICHAIN_F32_DIV1000(gravity, 64036, ICHAIN_STOP),
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:255]: yes, this file is like a year in the making lol

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 10]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+
+    if (limbIndex == STALFOS_LIMB_HEAD_ROOT) {
+        rot->x += this->headRot.y;
+        rot->y -= this->headRot.x;
+        rot->z += this->headRot.z;
+    } else if (limbIndex == STALFOS_LIMB_HEAD) {
+        OPEN_DISPS(globalCtx->state.gfxCtx, "../z_en_test.c", 3582);
+
+        gDPPipeSync(POLY_OPA_DISP++);
+        gDPSetEnvColor(POLY_OPA_DISP++, 80 + ABS((s16)(Math_SinS(globalCtx->gameplayFrames * 2000) * 175.0f)), 0, 0,
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:1862]: we dont enforce hex on all angles. if its a nicer number as decimal it makes more sense to write it that way, imo

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 11]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+        }
+
+        Actor_Kill(&this->actor);
+    }
+}
+
+void EnTest_UpdateHeadRot(EnTest* this, GlobalContext* globalCtx) {
+    s16 lookAngle = this->actor.yawTowardsPlayer;
+
+    lookAngle -= (s16)(this->headRot.y + this->actor.shape.rot.y);
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:1674]: cast is needed. it gets promoted to s32 during the addition, and the code has a deliberate cast. its a very common pattern

File: src/code/z_actor.c [hunk 12]
```diff
@@ -3388,24 +3388,33 @@ void Actor_SetTextWithPrefix(GlobalContext* globalCtx, Actor* actor, s16 baseTex
...
+    Vec3f curPos;
+
+    Math_Vec3f_Copy(&curPos, &actor->world.pos);
+    curBgCheckFlags = actor->bgCheckFlags;
+
+    newX = Math_SinS(yaw) * distance;
+    newZ = Math_CosS(yaw) * distance;
+
+    actor->world.pos.x += newX;
+    actor->world.pos.z += newZ;
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/code/z_actor.c:3411]: ill just revert the changes in my pr and let yours take care of it

[the remark this answers is not in the document]
Comment (fig02) [reply src/code/z_actor.c:3411]: if youd like, you can add the function level comment that I have here:
```
/**
 * Checks if a given actor will be standing on the ground after being translated
 * by the provided distance and angle.
 *
 * Returns true if the actor will be standing on ground.
 */
 ```

File: include/z64.h [hunk 13]
```diff
@@ -240,7 +240,7 @@ typedef struct {
 } ActorListEntry; // size = 0x08
 
 typedef struct {
-    /* 0x0000 */ u8     freezeFlashTimer;
+    /* 0x0000 */ u8     freezeFlashTimer; // on frames 4 through 1, the game will freeze and the screen will flash white
```
[the remark this answers is not in the document]
Comment (fig02) [reply include/z64.h:243]: tbh this doesnt really belong here, just gonna remove. The user should always accesses this through the function, which will set it to the appropriate values (and it also explains how it works there)

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 14]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+                EnTest_ChooseAction(this, globalCtx);
+            } else {
+                EnTest_SetupSidestepAgro(this, globalCtx);
+            }
+        }
+    }
+}
+
+// setup damaged
+void func_80862154(EnTest* this) {
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:1221]: Any reason for not just naming this?
Same for `func_808621D4`

**Review by louist103 — CHANGES_REQUESTED** [review]
Great job on this.  Just a few things
:StalZone:

File: src/overlays/actors/ovl_En_Test/z_en_test.h [hunk 15]
```diff
@@ -6,17 +6,110 @@
...
-    /* 0x01CC */ Vec3s jointTable_1CC[61];
-    /* 0x033A */ Vec3s transitionTbl_16E[61];
-    /* 0x04A8 */ SkelAnime Animation_4A8;
-    /* 0x04EC */ Vec3s jointTable_4EC[61];
-    /* 0x065A */ Vec3s transitionTbl_65A[61];
-    /* 0x07C8 */ char unk_7C8[0x160];
-} EnTest; // size = 0x0928
+    /* 0x000 */ Actor actor;
+    /* 0x14C */ Vec3s bodyPartsPos[10];
+    /* 0x188 */ SkelAnime skelAnime;
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.h:79]: no, the main skelanime is both upper and lower. the second skelanime will take over for the upper body when specified to

File: src/overlays/actors/ovl_En_Test/z_en_test.c (hunk 14 above)
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:1221]: the comments have since been removed. there are 3 different functions that handle different kinds of damage, so they need to be figured out

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 16]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+            case 4:
+            case 9:
+                if (this->actor.params != STALFOS_TYPE_CEILING) {
+                    this->actor.world.rot.y = this->actor.yawTowardsPlayer;
+                    EnTest_SetupJumpBack(this);
+                }
+                break;
+        }
+    } else if (yawDiff <= 0x3E80) {
+        if (ABS((s16)(this->actor.yawTowardsPlayer - this->actor.shape.rot.y)) > 0x3E80) {
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:399]: yes its needed, see the response to dragorn above on a similar cast. it gets promoted to s32 during the addition/subtraction

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 17]
```diff
@@ -133,126 +204,1885 @@ static ColliderQuadInit D_808645C8 = {
...
+            if (Actor_IsTargeted(globalCtx, &this->actor)) {
+                if (Rand_ZeroOne() < 0.1f) {
+                    EnTest_SetupJumpslash(this);
+                    return;
+                }
+            } else if (player->heldItemActionParam != PLAYER_AP_NONE) {
+                if (this->actor.isTargeted) {
+                    if ((globalCtx->gameplayFrames % 2) != 0) {
+                        EnTest_SetupSidestepAgro(this, globalCtx);
+                        return;
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:678]: you cant, the early return is needed so the rest of the function doesnt run

File: src/overlays/actors/ovl_En_Test/z_en_test.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (engineer124) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:399]: Just a pass-by observation, but this subtraction of two angles followed by an (s16) cast is common enough, especially in Camera, that there's already a macro `BINANG_SUB()` if you want to use it:
https://github.com/zeldaret/oot/blob/e632b9a1200e0411bbb54c3213682f32665f821e/include/z64math.h#L99

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 18]
```diff
@@ -133,126 +204,1863 @@ static ColliderQuadInit D_808645C8 = {
...
+    if (!EnTest_ReactToProjectile(globalCtx, this)) {
+        yawDiff = this->actor.yawTowardsPlayer;
+        yawDiff -= this->actor.shape.rot.y;
+
+        if (yawDiff > 0) {
+            yawChange = (yawDiff / 42.0f) + 300.0f;
+            this->actor.shape.rot.y += (yawChange * 2);
+        } else {
+            yawChange = (yawDiff / 42.0f) - 300.0f;
+            this->actor.shape.rot.y += (yawChange * 2);
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:742]:
```suggestion
            this->actor.shape.rot.y += yawChange * 2;
        } else {
            yawChange = (yawDiff / 42.0f) - 300.0f;
            this->actor.shape.rot.y += yawChange * 2;
```

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 19]
```diff
@@ -133,126 +204,1863 @@ static ColliderQuadInit D_808645C8 = {
...
+
+        if ((this->actor.colChkInfo.damageEffect != STALFOS_DMGEFF_SLING) &&
+            (this->actor.colChkInfo.damageEffect != STALFOS_DMGEFF_FIREMAGIC)) {
+            this->lastDamageEffect = this->actor.colChkInfo.damageEffect;
+            if (this->swordState >= STALFOS_SWORD_ON) {
+                this->swordState = STALFOS_SWORD_OFF;
+            }
+            this->unk_7DC = player->unk_845;
+            this->actor.world.rot.y = this->actor.yawTowardsPlayer;
+            Actor_SetDropFlag(&this->actor, &this->bodyCollider.info, 0);
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:1681]:
```suggestion
            Actor_SetDropFlag(&this->actor, &this->bodyCollider.info, false);
```

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 20]
```diff
@@ -133,126 +204,1863 @@ static ColliderQuadInit D_808645C8 = {
...
+        }
+
+    } else if ((limbIndex == STALFOS_LIMB_SHIELD) && (this->unk_7DE != 0)) {
+        Matrix_MultVec3f(&D_80864670, &sp64);
+
+        this->shieldCollider.dim.pos.x = sp64.x;
+        this->shieldCollider.dim.pos.y = sp64.y;
+        this->shieldCollider.dim.pos.z = sp64.z;
+    } else {
+        Actor_SetFeetPos(&this->actor, limbIndex, 48, &D_80864658, 55, &D_80864658);
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:1905]:
```suggestion
        Actor_SetFeetPos(&this->actor, limbIndex, STALFOS_LIMB_FOOT_L, &D_80864658, STALFOS_LIMB_ANKLE_R, &D_80864658);
```

**Review by Dragorn421 — CHANGES_REQUESTED** [review]
<details>
<summary>Click for meme</summary>



</details>

<details>
<summary>function flow (text)</summary>

```
EnTest_Init
	-(not ceiling type)> EnTest_SetupWaitGround -> EnTest_WaitGround -(player Y near)> EnTest_Rise -> EnTest_SetupJumpBack
	-(ceiling type)> EnTest_SetupWaitAbove -> EnTest_WaitAbove -> EnTest_Fall -> EnTest_Land -> EnTest_SetupIdle

EnTest_SetupJumpBack -> EnTest_JumpBack
	-(EnTest_ReactToProjectile)> ?
	-(player near and facing)> EnTest_SetupSlash1
	-(player not too far and facing)> EnTest_SetupJumpslash
	-(else)> EnTest_SetupIdle

EnTest_ReactToProjectile
	-(hookshot)> EnTest_SetupJumpUp
	-(projectile near, from front)> EnTest_SetupStopAndBlock
	-(projectile near, not from behind)> EnTest_SetupJumpBack
	-(projectile near)> EnTest_SetupJumpUp
	-(projectile from front or behind)> EnTest_SetupSidestepSetSpeed
	-(projectile not from behind and wall in front)> EnTest_SetupJumpBack
	-(projectile not from behind and wall behind)> EnTest_SetupJumpUp

EnTest_SetupIdle -> EnTest_Idle
	-(EnTest_ReactToProjectile)> ?
	-(player hit from side?)> EnTest_SetupJumpBack / EnTest_SetupSidestepAggro
	-(facing player, medium range, random)> EnTest_SetupJumpslash / EnTest_SetupSidestepAggro
	-(facing player, random)> EnTest_SetupWalkAndBlock / EnTest_SetupSidestepAggro
	-(not facing player, random)> EnTest_SetupSidestepFromIdle / EnTest_ChooseAction

EnTest_SetupJumpslash -> EnTest_Jumpslash -> EnTest_SetupIdle

EnTest_SetupWalkAndBlock -> EnTest_WalkAndBlock
	-(player near, swinging?)> EnTest_SetupJumpBack / EnTest_SetupStopAndBlock
	-(midrange, facing player, targeted)> EnTest_SetupJumpslash
	-(targeted, player holding item?)> EnTest_SetupSidestepAggro / EnTest_ChooseAction
	-(player holding item)> EnTest_SetupSidestepInactive
	-(not facing player)> EnTest_SetupIdle
	-(player near, rand, not targeted)> EnTest_SetupSidestepAggro
	-(player near, rand)> EnTest_SetupSlash1 / EnTest_SetupStopAndBlock

EnTest_SetupSidestepFromIdle -> EnTest_SidestepFromIdle
	-(EnTest_ReactToProjectile)> ?
	-(facing player)> EnTest_SetupSidestepInactive / EnTest_ChooseAction / EnTest_SetupWalkAndBlock

EnTest_SetupSlash1 -> EnTest_Slash1
	-(random)> EnTest_SetupSlash2 -> EnTest_Slash2 -> EnTest_SetupSlash1
	-> EnTest_SetupSlash1End -> EnTest_Slash1End

EnTest_Slash1End
	-(hit player and not ceiling type)> EnTest_SetupJumpBack
	-(random)> EnTest_SetupIdle
	-(random and not ceiling type)> EnTest_SetupJumpBack
	-(conditionals that don't make sense)> EnTest_SetupJumpBack / EnTest_SetupSlash1 / EnTest_SetupSidestepAggro

EnTest_SetupJumpUp -> EnTest_JumpUp -> EnTest_Land -> EnTest_SetupIdle

EnTest_SetupStopAndBlock -> EnTest_StopAndBlock
	-(player not in front and not ceiling type)> EnTest_SetupJumpBack
	-(after ~10-20 frames)> EnTest_SetupIdleFromBlock

EnTest_SetupIdleFromBlock -> EnTest_IdleFromBlock
	-(EnTest_ReactToProjectile)> ?
	-(player far)> EnTest_SetupSidestepAggro
	-(else)> EnTest_ChooseAction

EnTest_SetupSidestepSetSpeed -> EnTest_SidestepAggro

EnTest_SetupSidestepAggro
	-(not targeted)> EnTest_SetupSidestepInactive
	-> EnTest_SidestepAggro

EnTest_SidestepAggro
	-> EnTest_SetupIdle
	-(targeted)> (EnTest_ReactToProjectile?) / EnTest_ChooseAction
	-> EnTest_SetupIdle / EnTest_SetupWalkAndBlock

EnTest_SetupSidestepInactive -> EnTest_SidestepInactive -> EnTest_ChooseAction

EnTest_ChooseAction
	-(player looks in opposite direction)> EnTest_SetupStopAndBlock / EnTest_SetupSidestepAggro / EnTest_SetupJumpBack
	-(player doesn't look in same direction)> EnTest_ChooseBackupAction
	-(player not in front, random, not ceiling type)> EnTest_SetupJumpBack
	-((dead code?) player not in front, middle range, facing player, not targeted)> EnTest_SetupJumpslash
	-(player not in ~front)> EnTest_SetupWalkAndBlock
	-(player in ~front)> EnTest_SetupSlash1 / EnTest_SetupSidestepAggro
	-> EnTest_ChooseBackupAction

EnTest_ChooseBackupAction
	-(midrange, facing player, targeted)> EnTest_SetupJumpslash
	-> EnTest_SetupWalkAndBlock / EnTest_SetupSidestepAggro / EnTest_SetupStopAndBlock

// incomplete:

EnTest_Update
	-> EnTest_UpdateDamage
		-(nut, boomerang, hookshot, ice, light)> EnTest_SetupStunned
		-(player in front, dead)> func_80862FA8
	-(sword bounced)> EnTest_SetupRecoil

EnTest_SetupStunned -> EnTest_Stunned

EnTest_SetupRecoil -> EnTest_Recoil
	-(random and not ceiling type)> EnTest_SetupJumpBack
	-> EnTest_SetupIdle / EnTest_SetupSidestepAggro

func_80862154
func_808621D4
func_80862398
func_80862418
func_80862DBC -> func_80862E6C
func_80863044
func_808630F0
func_8086318C

func_808633E8
```

</details>

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 21]
```diff
@@ -15,53 +16,123 @@ void EnTest_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
+void func_808633E8(EnTest* this, GlobalContext* globalCtx);
+void func_80862FA8(EnTest* this, GlobalContext* globalCtx);
+
+s32 EnTest_ReactToProjectile(GlobalContext* globalCtx, EnTest* this);
+
+typedef enum {
+    /* -1 */ STALFOS_SWORD_NOBLUR = -1,
+    /*  0 */ STALFOS_SWORD_OFF,
+    /*  1 */ STALFOS_SWORD_ON
+} StalfosSwordState;
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:70]: add space will disable the blur completely, which is why I said noblur. I dont think its a fade. I dont fully understand the purpose of the -1 value tho so I may still be missing something.

I think we will end up renaming Blure to Blur, and I would like to keep that consistent. Blur as a name makes sense to me, its the blur created by the motion.

We use the "better" names in enums and macros all over the place. Its fine

File: src/code/z_actor.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (Dragorn421) [reply src/code/z_actor.c]: done

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 22]
```diff
@@ -15,53 +16,123 @@ void EnTest_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
-void func_8086209C(EnTest* this, GlobalContext* globalCtx);
+void EnTest_SetupWaitGround(EnTest* this);
+void EnTest_SetupWaitAbove(EnTest* this);
+void EnTest_SetupJumpBack(EnTest* this);
+void EnTest_SetupSlash1End(EnTest* this);
+void EnTest_SetupSlash2(EnTest* this);
+void EnTest_SetupJumpslash(EnTest* this);
+void EnTest_SetupWalkAndBlock(EnTest* this);
+void EnTest_SetupSidestepInactive(EnTest* this);
+void EnTest_SetupSlash1(EnTest* this);
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:27]: i like these names, but im gonna specify "SlashDownEnd". The animation wouldnt look right if you tried to transition to end from the upward slash

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 23]
```diff
@@ -133,126 +204,1863 @@ static ColliderQuadInit D_808645C8 = {
...
+    }
+}
+
+void EnTest_SetupSidestepSetSpeed(EnTest* this, f32 xzSpeed) {
+    Animation_MorphToLoop(&this->skelAnime, &gStalfosSidestepAnim, -2.0f);
+    this->actor.speedXZ = xzSpeed;
+    this->actor.world.rot.y = this->actor.shape.rot.y + 0x3FFF;
+    this->timer = (Rand_ZeroOne() * 20.0f) + 15.0f;
+    this->unk_7C8 = 0x18;
+    EnTest_SetupAction(this, EnTest_SidestepAggro);
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:1987]: These names made sense to me at the time when I made them, but I honestly forgot alot of my reasoning behind them. I will probably revert the sidestep names to `func_` names for now, and myself or someone else can do a second documentation pass later

File: src/overlays/actors/ovl_En_Test/z_en_test.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (fig02) [reply src/overlays/actors/ovl_En_Test/z_en_test.c:70]: Ill be removing this enum for now, I still dont fully understand the purpose behind the existence of the -1 value. May make more sense after some more blure documentation

**Review by Dragorn421 — APPROVED** [review]
Apart from adding a `@bug` comment in `EnTest_ChooseAction` if I'm right about the dead code,
This looks good to me

File: assets/xml/objects/object_sk2.xml [hunk 24]
```diff
@@ -1,55 +1,130 @@
...
+        <DList Name="gStalfosLeftShoulderArmorDL" Offset="0x64B0" />
+        <DList Name="gStalfosLeftFootDL" Offset="0x6580" />
+        <DList Name="gStalfosLeftForearmDL" Offset="0x6680" />
+        <DList Name="gStalfosLeftLowerLegDL" Offset="0x6778" />
+        <DList Name="gStalfosLeftThighDL" Offset="0x68E0" />
+        <DList Name="gStalfosLeftHandDL" Offset="0x6988" />
+        <DList Name="gStalfosSwordDL" Offset="0x6A48" />
+
+        <!-- Stalfos textures -->
+        <Texture Name="gStalfosEyeSocketTex" OutName="stalfos_eye_socket" Format="rgb5a1" Width="16" Height="16" Offset="0x6BA8" />
```
Comment (Roman971) [inline assets/xml/objects/object_sk2.xml:30]: this and the others need to be updated to rgba16 ```suggestion
        <Texture Name="gStalfosEyeSocketTex" OutName="stalfos_eye_socket" Format="rgba16" Width="16" Height="16" Offset="0x6BA8" />
```

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 25]
```diff
@@ -133,126 +198,1867 @@ static ColliderQuadInit D_808645C8 = {
...
+        yawToProjectile = Actor_WorldYawTowardActor(&this->actor, projectileActor) - (u16)this->actor.shape.rot.y;
+
+        if ((u8)(this->actor.bgCheckFlags & 8)) {
+            wallYawDiff = ((u16)this->actor.wallYaw - (u16)this->actor.shape.rot.y);
+            touchingWall = true;
+        } else {
+            touchingWall = false;
+        }
+
+        if ((Math_Vec3f_DistXYZ(&this->actor.world.pos, &projectileActor->world.pos) < 200.0f)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:2011]:
```suggestion
        if (Math_Vec3f_DistXYZ(&this->actor.world.pos, &projectileActor->world.pos) < 200.0f) {
```

File: src/overlays/actors/ovl_En_Test/z_en_test.c [hunk 26]
```diff
@@ -133,126 +198,1867 @@ static ColliderQuadInit D_808645C8 = {
...
+                bodyPart = 5;
+                break;
+            case STALFOS_LIMB_WAIST:
+                bodyPart = 6;
+                break;
+            case STALFOS_LIMB_FOOT_L:
+                bodyPart = 7;
+                break;
+            case STALFOS_LIMB_FOOT_R:
+                bodyPart = 8;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Test/z_en_test.c:1941]:
```suggestion
                bodyPart = 8;
                break;
```

**Comment by fig02** [comment]
changes made

**Review by Roman971 — APPROVED** [review]
